File: code/ex198.py
#f(0) = nums[0]
#f(1) = max(nums[1], nums[0])
#f(2) = max(f(1), f(0) + nums(2))
#f(3) = max(f(2), f(1) + nums(3))
#...
#f(n) = max(f(n-1), f(n - 2) + nums(n))


class Solution:
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        #Pre-processing
        if len(nums) == 0:
            return 0

        if len(nums) == 1:
            return nums[0]


        #DP
        f = [0 for i in range(len(nums))]

        f[0] = nums[0]
        f[1] = max(f[0], nums[1])


        for i in range(2, len(nums)):
            f[i] = max(f[i - 1], f[i - 2] + nums[i])



        return f[-1]


sol = Solution()
print(sol.rob([2,7,9,3,1]))

File: code/ex198.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 0)    return 0;
        else if (n == 1)    return nums[0];
        else if (n == 2)    return max(nums[0], nums[1]);
        int dp11 = nums[0], dp1 = max(nums[0], nums[1]);
        for (int i = 2; i < n; i++) {
            int nxt = max(dp11 + nums[i], dp1);
            dp11 = dp1;
            dp1 = nxt;
        }
        return dp1;
    }
};